File: leetcode/21. Merge Two Sorted Lists/solution1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *ref, *ret;
        
        if (l1 == NULL) {
            return l2;
        }
        
        else if (l2 == NULL) {
            return l1;
        }
        
        if (l1 -> val < l2 -> val) {
            ref = l1;
            l1 = l1 -> next;
        }
        
        else {
            ref = l2;
            l2 = l2 -> next;
        }
        
        ret = ref;
        
        while (l1 != NULL || l2 != NULL) {
            if (l1 == NULL) {
                ref -> next = l2;
                l2 = l2 -> next;
                ref = ref -> next;
            }
            
            else if (l2 == NULL) {
                ref -> next = l1;
                l1 = l1 -> next;
                ref = ref -> next;
            }
            
            else if (l1 -> val <= l2 -> val) {
                ref -> next = l1;
                l1 = l1 -> next;
                ref = ref -> next;
            }
            else {
                ref -> next = l2;
                l2 = l2 -> next;
                ref = ref -> next;
            }
        }
        
        return ret;
    }
};

File: leetcode/2. Add Two Numbers/solution1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        carry = 0
        
        out = ListNode()
        ret = out
        
        while True:
            
            if l1 != None and l2 != None:
                curr_add = l1.val + l2.val + carry
                out.val = curr_add % 10
                carry = curr_add // 10
                
                l1 = l1.next
                l2 = l2.next
            
            elif l1 == None and l2 != None:
                curr_add = l2.val + carry
                out.val = curr_add % 10
                carry = curr_add // 10
                
                l2 = l2.next
            
            elif l2 == None and l1 != None:
                curr_add = l1.val + carry
                out.val = curr_add % 10
                carry = curr_add // 10
                
                l1 = l1.next
            
            if l1 == None and l2 == None:
                if carry == 0:
                    return ret
                else:
                    out.next = ListNode()
                    out = out.next
                    out.val = carry
                    return ret
                
            out.next = ListNode()
            out = out.next
